crates/telio-starcast/src/nat.rs

```diff
@@ -0,0 +1,648 @@
+use crate::utils::MutableIpPacket;
+use pnet_packet::{
+    ip::IpNextHeaderProtocols,
+    ipv4::{self, MutableIpv4Packet},
+    ipv6::MutableIpv6Packet,
+    udp::{self, MutableUdpPacket},
+};
+use std::{
+    hash::{DefaultHasher, Hash, Hasher},
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+    time::Duration,
+};
+use telio_utils::{telio_log_warn, LruCache};
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// The packet is too short to be a valid IP packet
+    PacketTooShort,
+    /// The packet has an unexpected IP version
+    UnexpectedIpVersion,
+    /// The packet has an unexpected transport protocol
+    UnexpectedTransportProtocol,
+    /// There is no mapping for the given port
+    MappingNotFound,
+}
+
+pub trait Nat {
+    /// Translate outgoing packet
+    /// Change the destination to the peer's original ip and port
+    fn translate_outgoing(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+    /// Translate incomming packet
+    /// Change the source to the Starcast's virtual ip and natted port
+    fn translate_incomming(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+}
+
+type Mapper = fn(src_addr: SocketAddr) -> u16;
+
+fn default_mapper(src_addr: SocketAddr) -> u16 {
+    let mut hasher = DefaultHasher::new();
+    match src_addr.ip() {
+        IpAddr::V4(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+        IpAddr::V6(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+    };
+    src_addr.port().hash(&mut hasher);
+    hasher.finish() as u16 // TODO: think if we want to use the full 16 bits or maybe some range e.g. >=1024
+}
+
+pub struct StarcastNat {
+    mappings: LruCache<u16, SocketAddr>,
+    vpeer_ipv4_addr: Ipv4Addr,
+    vpeer_ipv6_addr: Ipv6Addr,
+    mapper: Mapper,
+}
+
+impl StarcastNat {
+    /// Create new starcast nat with default capcity of 2048 entries and 2 minute timeout
+    /// 2min default is taken from (https://datatracker.ietf.org/doc/html/rfc4787#section-4.3)
+    pub fn new(vpeer_ipv4: Ipv4Addr, vpeer_ipv6: Ipv6Addr) -> Self {
+        const LRU_CAPACITY: usize = 2048;
+        const LRU_TIMEOUT: Duration = Duration::from_secs(120);
+        Self::new_custom(
+            vpeer_ipv4,
+            vpeer_ipv6,
+            LRU_CAPACITY,
+            LRU_TIMEOUT,
```

Comment:
I disagree. Because I don't want to expose what type of cache I'm using internally. Imagine now you have 100 tests using this new_custom method. If you want to then change the implementation and use different cache internally you need to modify all of those tests.

Reply:
Hmm, maybe? Maybe `DEFAULT_*` would be better for the names as every capacity is LRU capacity here, so there is nothing specially LRU about this one? idk